jupyterhub/tests/test_scopes.py

```diff
@@ -0,0 +1,195 @@
+"""Test scopes for API handlers"""
+import pytest
+from pytest import mark
+from tornado import web
+
+from .. import orm
+from ..utils import check_scope
+from ..utils import needs_scope
+from ..utils import parse_scopes
+from ..utils import Scope
+from .utils import add_user
+from .utils import api_request
+from .utils import auth_header
+
+
+def test_scope_constructor():
+    user1 = 'george'
+    user2 = 'michael'
+    scope_list = [
+        'users',
+        'read:users!user={}'.format(user1),
+        'read:users!user={}'.format(user2),
+    ]
+    parsed_scopes = parse_scopes(scope_list)
+
+    assert 'read:users' in parsed_scopes
+    assert parsed_scopes['users']
+    assert set(parsed_scopes['read:users']['user']) == {user1, user2}
+
+
+def test_scope_precendence():
+    scope_list = ['read:users!user=maeby', 'read:users']
+    parsed_scopes = parse_scopes(scope_list)
+    assert parsed_scopes['read:users'] == Scope.ALL
+
+
+def test_scope_check_present():
+    handler = None
+    scope_list = ['read:users']
+    parsed_scopes = parse_scopes(scope_list)
+    assert check_scope(handler, 'read:users', parsed_scopes)
+    assert check_scope(handler, 'read:users', parsed_scopes, user='maeby')
+
+
+def test_scope_check_not_present():
+    handler = None
+    scope_list = ['read:users!user=maeby']
+    parsed_scopes = parse_scopes(scope_list)
+    assert not check_scope(handler, 'read:users', parsed_scopes)
+    assert not check_scope(handler, 'read:users', parsed_scopes, user='gob')
+    assert not check_scope(
+        handler, 'read:users', parsed_scopes, user='gob', server='server'
+    )
+
+
+def test_scope_filters():
+    handler = None
+    scope_list = ['read:users', 'read:users!group=bluths', 'read:users!user=maeby']
+    parsed_scopes = parse_scopes(scope_list)
+    assert check_scope(handler, 'read:users', parsed_scopes, group='bluth')
+    assert check_scope(handler, 'read:users', parsed_scopes, user='maeby')
+
+
+def test_scope_one_filter_only():
+    handler = None
+    with pytest.raises(AttributeError):
+        check_scope(
+            handler, 'all', parse_scopes(['all']), user='george_michael', group='bluths'
+        )
+
+
+def test_scope_parse_server_name():
+    handler = None
+    scope_list = ['users:servers!server=maeby/server1', 'read:users!user=maeby']
+    parsed_scopes = parse_scopes(scope_list)
+    assert check_scope(
+        handler, 'users:servers', parsed_scopes, user='maeby', server='server1'
+    )
+
+
+class MockAPIHandler:
+    def __init__(self):
+        self.scopes = ['users']
```

Comment:
something like:

```python
self.request = Mock(spec=tornado.httputil.HTTPServerRequest)
```

or build the mock object yourself:

```python
self.request = Mock()
self.request.path = "/thing"
```
[ref](https://www.tornadoweb.org/en/stable/httputil.html#tornado.httputil.HTTPServerRequest)